File: fons/nanus-go/lexer.go
package main

import (
	"strings"

	"subsidia"
)

var keywords = map[string]struct{}{
	// Declarations
	"varia": {}, "fixum": {}, "figendum": {}, "variandum": {},
	"functio": {}, "genus": {}, "pactum": {}, "ordo": {}, "discretio": {}, "typus": {},
	"ex": {}, "importa": {}, "ut": {},
	// Modifiers
	"publica": {}, "privata": {}, "protecta": {}, "generis": {}, "implet": {}, "sub": {}, "abstractus": {},
	// Control flow
	"si": {}, "sin": {}, "secus": {}, "dum": {}, "fac": {}, "elige": {}, "casu": {}, "ceterum": {}, "discerne": {}, "custodi": {},
	"de": {}, "in": {}, "pro": {}, "omnia": {},
	// Actions
	"redde": {}, "reddit": {}, "rumpe": {}, "perge": {}, "iace": {}, "mori": {}, "tempta": {}, "cape": {}, "demum": {},
	"scribe": {}, "vide": {}, "mone": {}, "adfirma": {}, "tacet": {},
	// Expressions
	"cede": {}, "novum": {}, "clausura": {}, "qua": {}, "innatum": {}, "finge": {},
	"sic": {}, "scriptum": {},
	// Operators (word-form)
	"et": {}, "aut": {}, "vel": {}, "inter": {}, "intra": {},
	"non": {}, "nihil": {}, "nonnihil": {}, "positivum": {}, "negativum": {}, "nulla": {}, "nonnulla": {},
	// Conversion operators
	"numeratum": {}, "fractatum": {}, "textatum": {}, "bivalentum": {},
	// Literals
	"verum": {}, "falsum": {}, "ego": {},
	// Entry
	"incipit": {}, "incipiet": {},
	// Test
	"probandum": {}, "proba": {},
	// Type
	"usque": {},
	// Annotations
	"publicum": {}, "externa": {},
}

var punctuators = map[byte]struct{}{
	'(': {}, ')': {}, '{': {}, '}': {}, '[': {}, ']': {},
	',': {}, '.': {}, ':': {}, ';': {}, '@': {}, '#': {},
	'?': {}, '!': {},
}

var operators = []string{
	// Multi-char first (greedy match)
	"===", "!==", "==", "!=", "<=", ">=", "&&", "||", "??",
	"+=", "-=", "*=", "/=",
	"->", "..",
	// Single-char
	"+", "-", "*", "/", "%",
	"<", ">", "=",
	"&", "|", "^", "~",
}

// Lex converts source text into tokens.
func Lex(source string, filename string) []subsidia.Token {
	tokens := []subsidia.Token{}
	pos := 0
	linea := 1
	lineStart := 0

	locus := func() subsidia.Locus {
		return subsidia.Locus{Linea: linea, Columna: pos - lineStart + 1, Index: pos}
	}

	peek := func(offset int) byte {
		idx := pos + offset
		if idx >= len(source) {
			return 0
		}
		return source[idx]
	}

	advance := func() byte {
		ch := source[pos]
		pos++
		if ch == '\n' {
			linea++
			lineStart = pos
		}
		return ch
	}

	match := func(str string) bool {
		if strings.HasPrefix(source[pos:], str) {
			for i := 0; i < len(str); i++ {
				advance()
			}
			return true
		}
		return false
	}

	skipWhitespace := func() {
		for pos < len(source) {
			ch := peek(0)
			if ch == ' ' || ch == '\t' || ch == '\r' {
				advance()
			} else if ch == '\n' {
				loc := locus()
				advance()
				tokens = append(tokens, subsidia.Token{Tag: subsidia.TokenNewline, Valor: "\n", Locus: loc})
			} else {
				break
			}
		}
	}

	readString := func(quote byte) string {
		var b strings.Builder
		advance()
		for pos < len(source) && peek(0) != quote {
			if peek(0) == '\\' {
				advance()
				esc := advance()
				switch esc {
				case 'n':
					b.WriteByte('\n')
				case 't':
					b.WriteByte('\t')
				case 'r':
					b.WriteByte('\r')
				case '\\':
					b.WriteByte('\\')
				case '"':
					b.WriteByte('"')
				case '\'':
					b.WriteByte('\'')
				default:
					b.WriteByte(esc)
				}
			} else {
				b.WriteByte(advance())
			}
		}
		advance()
		return b.String()
	}

	readTripleString := func() string {
		advance()
		advance()
		advance()

		if peek(0) == '\n' {
			advance()
		}

		var b strings.Builder
		for pos < len(source) {
			if peek(0) == '"' && peek(1) == '"' && peek(2) == '"' {
				value := b.String()
				if strings.HasSuffix(value, "\n") {
					value = strings.TrimSuffix(value, "\n")
				}
				advance()
				advance()
				advance()
				return value
			}
			b.WriteByte(advance())
		}
		return b.String()
	}

	readNumber := func() string {
		var b strings.Builder
		for pos < len(source) && isNumberChar(peek(0)) {
			b.WriteByte(advance())
		}
		return b.String()
	}

	readIdentifier := func() string {
		var b strings.Builder
		for pos < len(source) && isIdentChar(peek(0)) {
			b.WriteByte(advance())
		}
		return b.String()
	}

	readComment := func() string {
		var b strings.Builder
		advance()
		for pos < len(source) && peek(0) != '\n' {
			b.WriteByte(advance())
		}
		return b.String()
	}

	for pos < len(source) {
		skipWhitespace()
		if pos >= len(source) {
			break
		}

		loc := locus()
		ch := peek(0)

		if ch == '#' {
			value := readComment()
			tokens = append(tokens, subsidia.Token{Tag: subsidia.TokenComment, Valor: value, Locus: loc})
			continue
		}

		if ch == '"' && peek(1) == '"' && peek(2) == '"' {
			value := readTripleString()
			tokens = append(tokens, subsidia.Token{Tag: subsidia.TokenTextus, Valor: value, Locus: loc})
			continue
		}
		if ch == '"' || ch == '\'' {
			value := readString(ch)
			tokens = append(tokens, subsidia.Token{Tag: subsidia.TokenTextus, Valor: value, Locus: loc})
			continue
		}

		if isDigit(ch) {
			value := readNumber()
			tokens = append(tokens, subsidia.Token{Tag: subsidia.TokenNumerus, Valor: value, Locus: loc})
			continue
		}

		if isAlpha(ch) || ch == '_' {
			value := readIdentifier()
			_, isKeyword := keywords[value]
			tag := subsidia.TokenIdentifier
			if isKeyword {
				tag = subsidia.TokenKeyword
			}
			tokens = append(tokens, subsidia.Token{Tag: tag, Valor: value, Locus: loc})
			continue
		}

		matched := false
		for _, op := range operators {
			if match(op) {
				tokens = append(tokens, subsidia.Token{Tag: subsidia.TokenOperator, Valor: op, Locus: loc})
				matched = true
				break
			}
		}
		if matched {
			continue
		}

		// § is multi-byte UTF-8, handle via string match
		if match("§") {
			tokens = append(tokens, subsidia.Token{Tag: subsidia.TokenPunctuator, Valor: "§", Locus: loc})
			continue
		}

		if _, ok := punctuators[ch]; ok {
			advance()
			tokens = append(tokens, subsidia.Token{Tag: subsidia.TokenPunctuator, Valor: string(ch), Locus: loc})
			continue
		}

		panic(&subsidia.CompileError{Message: "unexpected character '" + string(ch) + "'", Locus: loc, Filename: filename})
	}

	tokens = append(tokens, subsidia.Token{Tag: subsidia.TokenEOF, Valor: "", Locus: locus()})
	return tokens
}

func isDigit(ch byte) bool {
	return ch >= '0' && ch <= '9'
}

func isAlpha(ch byte) bool {
	return (ch >= 'a' && ch <= 'z') || (ch >= 'A' && ch <= 'Z')
}

func isIdentChar(ch byte) bool {
	return isAlpha(ch) || isDigit(ch) || ch == '_'
}

func isNumberChar(ch byte) bool {
	return isDigit(ch) || ch == '.' || ch == '_'
}

File: fons/nanus-rs/src/lexer.rs
use std::collections::HashSet;
use subsidia_rs::{CompileError, Locus, Token};

/// Tokenize Faber source code.
pub fn lex(source: &str, filename: &str) -> Result<Vec<Token>, CompileError> {
    let mut lexer = Lexer::new(source, filename);
    lexer.lex()
}

struct Lexer<'a> {
    source: &'a str,
    bytes: &'a [u8],
    pos: usize,
    linea: i32,
    line_start: usize,
    filename: String,
    keywords: HashSet<&'static str>,
    punctuators: HashSet<char>,
    operators: Vec<&'static str>,
}

impl<'a> Lexer<'a> {
    fn new(source: &'a str, filename: &str) -> Self {
        let keywords: HashSet<&str> = [
            // Declarations
            "varia",
            "fixum",
            "functio",
            "genus",
            "pactum",
            "ordo",
            "discretio",
            "ex",
            "importa",
            "ut",
            // Modifiers
            "publica",
            "privata",
            "protecta",
            "generis",
            "implet",
            "sub",
            // Control flow
            "si",
            "sin",
            "secus",
            "dum",
            "fac",
            "elige",
            "casu",
            "ceterum",
            "discerne",
            "custodi",
            "de",
            "itera",
            "in",
            "pro",
            "omnia",
            // Actions
            "redde",
            "reddit",
            "rumpe",
            "perge",
            "iace",
            "mori",
            "tempta",
            "cape",
            "demum",
            "scribe",
            "vide",
            "mone",
            "adfirma",
            "tacet",
            // Expressions
            "cede",
            "novum",
            "clausura",
            "qua",
            "innatum",
            "finge",
            "sic",
            "scriptum",
            // Operators (word-form)
            "et",
            "aut",
            "vel",
            "inter",
            "intra",
            "non",
            "nihil",
            "nonnihil",
            "positivum",
            // Literals
            "verum",
            "falsum",
            "ego",
            // Entry
            "incipit",
            "incipiet",
            // Test
            "probandum",
            "proba",
            // Type
            "usque",
            // Annotations
            "publicum",
            "externa",
            // Body shortcuts
            "ergo",
            "tacet",
            "iacit",
            "moritor",
            // Iteration
            "ceteri",
        ]
        .into_iter()
        .collect();

        let punctuators: HashSet<char> = [
            '(', ')', '{', '}', '[', ']', ',', '.', ':', ';', '@', '#', '?', '!',
        ]
        .into_iter()
        .collect();

        let operators = vec![
            // Multi-char first (greedy match)
            "===", "!==", "==", "!=", "<=", ">=", "&&", "||", "??", "+=", "-=", "*=", "/=", "->",
            "..", // Single-char
            "+", "-", "*", "/", "%", "<", ">", "=", "&", "|", "^", "~",
        ];

        Self {
            source,
            bytes: source.as_bytes(),
            pos: 0,
            linea: 1,
            line_start: 0,
            filename: filename.to_string(),
            keywords,
            punctuators,
            operators,
        }
    }

    fn locus(&self) -> Locus {
        Locus {
            linea: self.linea,
            columna: (self.pos - self.line_start + 1) as i32,
            index: self.pos as i32,
        }
    }

    fn peek(&self, offset: usize) -> u8 {
        let idx = self.pos + offset;
        if idx >= self.bytes.len() {
            0
        } else {
            self.bytes[idx]
        }
    }

    fn advance(&mut self) -> u8 {
        let ch = self.bytes[self.pos];
        self.pos += 1;
        if ch == b'\n' {
            self.linea += 1;
            self.line_start = self.pos;
        }
        ch
    }

    fn match_str(&mut self, s: &str) -> bool {
        if self.source[self.pos..].starts_with(s) {
            for _ in 0..s.len() {
                self.advance();
            }
            true
        } else {
            false
        }
    }

    fn lex(&mut self) -> Result<Vec<Token>, CompileError> {
        let mut tokens = Vec::new();

        while self.pos < self.bytes.len() {
            self.skip_whitespace(&mut tokens);
            if self.pos >= self.bytes.len() {
                break;
            }

            let loc = self.locus();
            let ch = self.peek(0);

            // Comment
            if ch == b'#' {
                let value = self.read_comment();
                tokens.push(Token {
                    tag: "Comment".to_string(),
                    valor: value,
                    locus: loc,
                });
                continue;
            }

            // Triple-quoted string
            if ch == b'"' && self.peek(1) == b'"' && self.peek(2) == b'"' {
                let value = self.read_triple_string();
                tokens.push(Token {
                    tag: "Textus".to_string(),
                    valor: value,
                    locus: loc,
                });
                continue;
            }

            // String
            if ch == b'"' || ch == b'\'' {
                let value = self.read_string(ch);
                tokens.push(Token {
                    tag: "Textus".to_string(),
                    valor: value,
                    locus: loc,
                });
                continue;
            }

            // Number
            if is_digit(ch) {
                let value = self.read_number();
                tokens.push(Token {
                    tag: "Numerus".to_string(),
                    valor: value,
                    locus: loc,
                });
                continue;
            }

            // Identifier or keyword
            if is_alpha(ch) || ch == b'_' {
                let value = self.read_identifier();
                let tag = if self.keywords.contains(value.as_str()) {
                    "Keyword"
                } else {
                    "Identifier"
                };
                tokens.push(Token {
                    tag: tag.to_string(),
                    valor: value,
                    locus: loc,
                });
                continue;
            }

            // Operators (greedy match)
            let mut matched = false;
            for op in &self.operators.clone() {
                if self.match_str(op) {
                    tokens.push(Token {
                        tag: "Operator".to_string(),
                        valor: op.to_string(),
                        locus: loc,
                    });
                    matched = true;
                    break;
                }
            }
            if matched {
                continue;
            }

            // Section marker (UTF-8 multi-byte)
            if self.match_str("§") {
                tokens.push(Token {
                    tag: "Punctuator".to_string(),
                    valor: "§".to_string(),
                    locus: loc,
                });
                continue;
            }

            // Punctuators
            if self.punctuators.contains(&(ch as char)) {
                self.advance();
                tokens.push(Token {
                    tag: "Punctuator".to_string(),
                    valor: (ch as char).to_string(),
                    locus: loc,
                });
                continue;
            }

            return Err(CompileError::new(
                format!("unexpected character '{}'", ch as char),
                loc,
                &self.filename,
            ));
        }

        tokens.push(Token {
            tag: "EOF".to_string(),
            valor: String::new(),
            locus: self.locus(),
        });

        Ok(tokens)
    }

    fn skip_whitespace(&mut self, tokens: &mut Vec<Token>) {
        while self.pos < self.bytes.len() {
            let ch = self.peek(0);
            if ch == b' ' || ch == b'\t' || ch == b'\r' {
                self.advance();
            } else if ch == b'\n' {
                let loc = self.locus();
                self.advance();
                tokens.push(Token {
                    tag: "Newline".to_string(),
                    valor: "\n".to_string(),
                    locus: loc,
                });
            } else {
                break;
            }
        }
    }

    fn read_string(&mut self, quote: u8) -> String {
        let mut result = String::new();
        self.advance(); // skip opening quote

        while self.pos < self.bytes.len() && self.peek(0) != quote {
            if self.peek(0) == b'\\' {
                self.advance();
                let esc = self.advance();
                match esc {
                    b'n' => result.push('\n'),
                    b't' => result.push('\t'),
                    b'r' => result.push('\r'),
                    b'\\' => result.push('\\'),
                    b'"' => result.push('"'),
                    b'\'' => result.push('\''),
                    _ => result.push(esc as char),
                }
            } else {
                // Handle UTF-8 properly by reading chars from source slice
                let remaining = &self.source[self.pos..];
                if let Some(ch) = remaining.chars().next() {
                    result.push(ch);
                    let byte_len = ch.len_utf8();
                    for _ in 0..byte_len {
                        self.advance();
                    }
                } else {
                    self.advance();
                }
            }
        }
        if self.pos < self.bytes.len() {
            self.advance(); // skip closing quote
        }
        result
    }

    fn read_triple_string(&mut self) -> String {
        // Skip opening """
        self.advance();
        self.advance();
        self.advance();

        // Skip leading newline if present
        if self.peek(0) == b'\n' {
            self.advance();
        }

        let mut result = String::new();
        while self.pos < self.bytes.len() {
            if self.peek(0) == b'"' && self.peek(1) == b'"' && self.peek(2) == b'"' {
                // Trim trailing newline
                if result.ends_with('\n') {
                    result.pop();
                }
                self.advance();
                self.advance();
                self.advance();
                return result;
            }
            // Handle UTF-8 properly
            let remaining = &self.source[self.pos..];
            if let Some(ch) = remaining.chars().next() {
                result.push(ch);
                let byte_len = ch.len_utf8();
                for _ in 0..byte_len {
                    self.advance();
                }
            } else {
                self.advance();
            }
        }
        result
    }

    fn read_number(&mut self) -> String {
        let mut result = String::new();
        while self.pos < self.bytes.len() && is_number_char(self.peek(0)) {
            result.push(self.advance() as char);
        }
        result
    }

    fn read_identifier(&mut self) -> String {
        let mut result = String::new();
        while self.pos < self.bytes.len() && is_ident_char(self.peek(0)) {
            result.push(self.advance() as char);
        }
        result
    }

    fn read_comment(&mut self) -> String {
        let mut result = String::new();
        self.advance(); // skip #
        while self.pos < self.bytes.len() && self.peek(0) != b'\n' {
            result.push(self.advance() as char);
        }
        result
    }
}

fn is_digit(ch: u8) -> bool {
    ch >= b'0' && ch <= b'9'
}

fn is_alpha(ch: u8) -> bool {
    (ch >= b'a' && ch <= b'z') || (ch >= b'A' && ch <= b'Z')
}

fn is_ident_char(ch: u8) -> bool {
    is_alpha(ch) || is_digit(ch) || ch == b'_'
}

fn is_number_char(ch: u8) -> bool {
    is_digit(ch) || ch == b'.' || ch == b'_'
}
